File: runner-ios/FleetRunner/Protocol.swift
import Foundation

// Swift mirror of collector/schemas ("schema": 1). Shared protocol, not
// shared code — the Android runner mirrors these independently in Kotlin.
// JSONEncoder/Decoder snake_case strategies map jobId <-> job_id etc.

struct ModelRef: Codable {
    let name: String
    let format: String
    let quant: String?
    let sha256: String
}

struct Targets: Codable {
    let pool: String?
    let match: String?
    let exclusive: Bool?
}

struct BenchParams: Codable {
    let promptTokens: Int?
    let genTokens: Int?
    let warmupIters: Int?
    let measureIters: Int?
    let nThreads: Int?
    /// Decoded and never read on iOS — the Android BenchmarkEngine turns a
    /// `benchmark` into a timed loop with it, this runner never has. Kept so
    /// the two protocols stay one protocol. Anything that wants a sustained
    /// run here should send the `thermal` workload, which is that idea done
    /// properly: a row per iteration rather than a mean over a warming device.
    let sustainedMinutes: Int?
    /// thermal: how long to keep measuring, in seconds. The workload's whole
    /// shape — it runs measured iterations until this elapses rather than a
    /// fixed count of them, because the answer is a curve against wall time.
    let durationS: Int?
    // batch / vision-eval / embed-eval / pipeline
    let inputSha256: String?
    let maxTokens: Int?
    let maxItems: Int?
    let computeUnits: String?
    let topic: String?
    let maxEvents: Int?
    let after: Int?
    /// embed-eval: the per-document token budget. Not the benchmark's prompt
    /// window — a document longer than this is truncated, identically on every
    /// device in the fleet, which is what keeps the recall numbers comparable.
    let nCtx: Int?
    // vantage
    /// The URLs to fetch. One result row per URL per repetition.
    let urls: [String]?
    /// How many times to sweep the whole list. Repetition-major, so the
    /// repeats of one URL are spread across the run.
    let repeats: Int?
    /// Per-request ceiling, so one hanging host cannot consume the job.
    let timeoutS: Int?
}

struct Constraints: Codable {
    let requireCharging: Bool?
    let minBatteryPct: Int?
}

struct JobSpec: Codable {
    let schema: Int
    let jobId: String
    let workload: String
    let executor: String
    let model: ModelRef?
    let backend: String?
    let params: BenchParams?
    let targets: Targets?
    let constraints: Constraints?
}

struct DeviceDescriptor: Codable {
    let model: String
    let soc: String
    let ramMb: Int64
    let os: String
    let appVer: String
}

struct Metrics: Codable {
    var loadMs: Int64?
    var prefillTokS: Double?
    var decodeTokS: Double?
    var ttftMs: Double?
    var peakMemMb: Int64?
    var memMethod: String?
    var thermal: [String]?
    var batteryStartPct: Int?
    var batteryEndPct: Int?

    // vision-eval. These used to ride in the LLM slots above -- accuracy in
    // decodeTokS, latency in ttftMs, throughput in prefillTokS -- and top-5 and
    // p95 had nowhere to go at all, so they reached only the uploaded report
    // artifact and never the results table. convertToSnakeCase maps these to
    // top1_pct, top5_pct, p50_ms, p95_ms and images_per_s.
    var top1Pct: Double?
    var top5Pct: Double?
    var p50Ms: Double?
    var p95Ms: Double?
    var imagesPerS: Double?

    // thermal. A benchmark row is a summary; a thermal row is a point on a
    // curve, and these two are what place it. convertToSnakeCase maps them to
    // elapsed_s and thermal_state, which is how fleet-collector's
    // schemas/metrics.json names them.
    //
    // `thermalState` is the single state at this sample, deliberately not the
    // `thermal` array above: the array is a bag of states a run passed
    // through, which cannot say when. A string rather than a number because
    // Android and iOS name their levels differently and flattening them would
    // invent a scale neither vendor publishes.
    var elapsedS: Double?
    var thermalState: String?

    // embed-eval. A query counts at k when at least one of its relevant
    // documents lands in its top k — the definition the collector's schema
    // pins for recall_at_1 ("fraction of queries whose top hit is relevant"),
    // which the wider ks have to share or the three numbers would not belong
    // to one series. `dim` rides along because recall is only comparable
    // within one embedding width: 0.71 from a 384-d model and 0.71 from a
    // 1024-d one are not the same claim. convertToSnakeCase maps these to
    // recall_at_1, recall_at_5, recall_at_10, docs_per_s and dim.
    var recallAt1: Double?
    var recallAt5: Double?
    var recallAt10: Double?
    var docsPerS: Double?
    var dim: Int?

    // vantage. One row per URL per repetition, carrying this device's view of
    // one request. Every phase is optional and that is load-bearing: a request
    // served over a pooled connection resolves no name, opens no socket and
    // shakes no hands, and reporting those as 0 would read as "instant"
    // rather than as "did not happen".
    //
    // `loadMs` above carries the whole transfer here rather than a model load
    // — the only name metrics.json offers for "how long until it was all
    // here". `networkType` is the point of the workload: the same URL measured
    // from cellular and from fibre is two answers, and a latency figure with
    // no idea what carried it is not comparable to anything.
    var dnsMs: Double?
    var connectMs: Double?
    var tlsMs: Double?
    var ttfbMs: Double?
    var networkType: String?

    /// Spelled out because `.convertToSnakeCase` cannot produce three of these
    /// names. The strategy splits on capitals and a digit is not one, so
    /// `recallAt1` encodes as **`recall_at1`** — one underscore short of the
    /// name metrics.json declares, and a metric under a name the collector
    /// does not know is a metric that silently never arrives. Every other case
    /// is listed with no raw value and still goes through the strategy, which
    /// leaves an already-snake_cased raw value alone; that is why `ttftMs` and
    /// friends keep working without being spelled out here.
    enum CodingKeys: String, CodingKey {
        case loadMs, prefillTokS, decodeTokS, ttftMs, peakMemMb, memMethod, thermal
        case batteryStartPct, batteryEndPct
        case top1Pct, top5Pct, p50Ms, p95Ms, imagesPerS
        case elapsedS, thermalState
        case recallAt1 = "recall_at_1"
        case recallAt5 = "recall_at_5"
        case recallAt10 = "recall_at_10"
        case docsPerS, dim
        case dnsMs, connectMs, tlsMs, ttfbMs, networkType
    }
}

struct BeaconSample: Codable {
    let batteryPct: Int
    let charging: Bool
    let thermal: String
}

struct ResultPost: Codable {
    var schema: Int = 1
    var kind: String
    var jobId: String?
    var deviceId: String
    var iter: Int?
    var final: Bool?
    var ok: Bool?
    var device: DeviceDescriptor?
    var metrics: Metrics?
    var beacon: BeaconSample?
    var error: String?
    var artifacts: [String]?
}

struct RegisterPost: Codable {
    let deviceId: String
    let descriptor: DeviceDescriptor
    let pools: [String]
    /// The workloads this runner can actually dispatch, so the queue only
    /// offers it work it can run. See FleetAgent.dispatchedWorkloads.
    let capabilities: [String]
}

enum FleetJSON {
    static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        return e
    }()
    static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()
}

File: runner-android/app/src/main/java/com/taylab/fleetrunner/protocol/Protocol.kt
package com.taylab.fleetrunner.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive

/**
 * Kotlin mirror of the collector's schemas/{job,result}.schema.json ("schema": 1).
 * Shared protocol, not shared code: the iOS runner mirrors these independently.
 */
val FleetJson = Json {
    ignoreUnknownKeys = true
    // encodeDefaults so `schema = 1` is always on the wire (the collector
    // rejects rows without it); explicitNulls=false keeps unset fields out.
    encodeDefaults = true
    explicitNulls = false
}

@Serializable
data class ModelRef(
    val name: String,
    val format: String,
    val quant: String? = null,
    val sha256: String,
)

@Serializable
data class Targets(
    val pool: String? = null,
    val match: String? = null,
    val exclusive: Boolean? = null,
)

@Serializable
data class Constraints(
    @SerialName("require_charging") val requireCharging: Boolean? = null,
    @SerialName("min_battery_pct") val minBatteryPct: Int? = null,
)

@Serializable
data class JobSpec(
    val schema: Int,
    @SerialName("job_id") val jobId: String,
    val workload: String,
    val executor: String,
    val model: ModelRef? = null,
    val backend: String? = null,
    val params: JsonObject? = null,
    val targets: Targets? = null,
    val constraints: Constraints? = null,
)

fun JsonObject?.intParam(key: String, default: Int): Int =
    this?.get(key)?.jsonPrimitive?.intOrNull ?: default

fun JsonObject?.stringParam(key: String): String? =
    this?.get(key)?.jsonPrimitive?.contentOrNull

/**
 * A params array of strings — `params.urls` for vantage. Missing reads as
 * empty rather than as an error, so the workload owns the "you asked for no
 * URLs" message instead of the JSON layer throwing something less useful.
 */
fun JsonObject?.stringListParam(key: String): List<String> =
    (this?.get(key) as? JsonArray)?.mapNotNull { it.jsonPrimitive.contentOrNull } ?: emptyList()

@Serializable
data class DeviceDescriptor(
    val model: String,
    val soc: String,
    @SerialName("ram_mb") val ramMb: Long,
    val os: String,
    @SerialName("app_ver") val appVer: String,
)

@Serializable
data class Metrics(
    @SerialName("load_ms") val loadMs: Long? = null,
    @SerialName("prefill_tok_s") val prefillTokS: Double? = null,
    @SerialName("decode_tok_s") val decodeTokS: Double? = null,
    @SerialName("ttft_ms") val ttftMs: Double? = null,
    @SerialName("peak_mem_mb") val peakMemMb: Long? = null,
    @SerialName("mem_method") val memMethod: String? = null,
    val thermal: List<String>? = null,
    @SerialName("battery_start_pct") val batteryStartPct: Int? = null,
    @SerialName("battery_end_pct") val batteryEndPct: Int? = null,

    // vision-eval. These used to ride in the LLM slots above -- accuracy in
    // decode_tok_s, latency in ttft_ms, throughput in prefill_tok_s -- and
    // top-5 and p95 had nowhere to go at all, so they only ever reached the
    // uploaded report artifact and never the results table.
    @SerialName("top1_pct") val top1Pct: Double? = null,
    @SerialName("top5_pct") val top5Pct: Double? = null,
    @SerialName("p50_ms") val p50Ms: Double? = null,
    @SerialName("p95_ms") val p95Ms: Double? = null,
    @SerialName("images_per_s") val imagesPerS: Double? = null,

    // thermal. The `thermal` array above is the whole run's state sequence;
    // these two describe one sample, which is what makes a row a point on a
    // curve rather than a summary of one. thermal_state is a String and not an
    // enum on purpose: Android and iOS name their levels differently, and
    // flattening them here would invent a scale neither vendor publishes.
    @SerialName("elapsed_s") val elapsedS: Double? = null,
    @SerialName("thermal_state") val thermalState: String? = null,

    // embed-eval. A query counts at k when at least one of its relevant docs
    // is in its top k, which is how the collector's schema defines
    // recall_at_1 ("fraction of queries whose top hit is relevant") and so is
    // what the other two have to mean as well. `dim` rides along because
    // recall is only comparable within one embedding width -- two models at
    // 0.71 mean nothing to each other if one of them is 384-d and the other
    // 1024-d. p50_ms/p95_ms are per-embedding latency, the same quantity the
    // vision eval puts there (per-item inference), over this run's embeddings.
    @SerialName("recall_at_1") val recallAt1: Double? = null,
    @SerialName("recall_at_5") val recallAt5: Double? = null,
    @SerialName("recall_at_10") val recallAt10: Double? = null,
    @SerialName("docs_per_s") val docsPerS: Double? = null,
    val dim: Int? = null,

    // vantage. One row per URL per repetition carries this device's view of
    // one request. Every phase is nullable on purpose: a request that reuses a
    // pooled connection resolves no name, opens no socket and shakes no hands,
    // and reporting those as 0 ms would read as "instant" rather than as "did
    // not happen". `load_ms` here is the whole transfer, not a model load --
    // the same field name the benchmark uses for its own load, which is the
    // only name metrics.json offers for "how long until it was all here".
    // `network_type` is what the fleet is actually for: the same URL measured
    // from cellular and from fibre is two answers, and a latency figure with
    // no idea what carried it is not comparable to anything.
    @SerialName("dns_ms") val dnsMs: Double? = null,
    @SerialName("connect_ms") val connectMs: Double? = null,
    @SerialName("tls_ms") val tlsMs: Double? = null,
    @SerialName("ttfb_ms") val ttfbMs: Double? = null,
    @SerialName("network_type") val networkType: String? = null,
)

@Serializable
data class BeaconSample(
    @SerialName("battery_pct") val batteryPct: Int,
    val charging: Boolean,
    val thermal: String,
)

@Serializable
data class ResultPost(
    val schema: Int = 1,
    val kind: String,
    @SerialName("job_id") val jobId: String? = null,
    @SerialName("device_id") val deviceId: String,
    val iter: Int? = null,
    val final: Boolean? = null,
    val ok: Boolean? = null,
    val device: DeviceDescriptor? = null,
    val metrics: Metrics? = null,
    val beacon: BeaconSample? = null,
    val error: String? = null,
    /** sha256 refs of output artifacts (batch results, reports). */
    val artifacts: List<String>? = null,
)

@Serializable
data class RegisterPost(
    @SerialName("device_id") val deviceId: String,
    val descriptor: DeviceDescriptor,
    val pools: List<String>,
    /** Workloads this runner can actually dispatch; the queue routes on them. */
    val capabilities: List<String> = emptyList(),
)
